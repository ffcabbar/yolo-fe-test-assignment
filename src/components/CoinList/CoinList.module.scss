.wrapper {
  display: flex;
  width: 20%;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid transparent;
  border-image: linear-gradient(50deg, gray 1%, rgba(251, 247, 252, 0) 100%);
  border-image-slice: 1;
  padding-bottom: 0.7rem;
  transition-duration: 0.2s;
  transition-timing-function: cubic-bezier(0.2, 0.8, 0.4, 1);
  margin-bottom: 1.5rem;

  @media (max-width: 1400px) {
    width: 24%;
  }

  @media (max-width: 992px) {
    width: 40%;
  }

  @media (max-width: 768px) {
    width: 100%;
  }

  &:hover {
    transform: translateY(-2px);
  }

  .leftSection {
    display: flex;
    align-items: center;
    gap: 2rem;
  }

  .price {
    color: white;
    display: grid;
    gap: 0.3rem;

    div {
      &:first-child {
        font-size: 1.2rem;
        font-weight: 500;

        @media (max-width: 1400px) {
          font-size: 1rem;
        }
      }
      &:last-child {
        font-size: 1rem;
        color: gray;
        font-weight: 500;

        @media (max-width: 1400px) {
          font-size: 0.8rem;
        }
      }
    }
  }

  .delete {
    cursor: pointer;
    margin-right: 1rem;
  }
}

.loading {
  display: flex;
  justify-content: flex-start;

  @media (max-width: 768px) {
    justify-content: center;
  }
}
