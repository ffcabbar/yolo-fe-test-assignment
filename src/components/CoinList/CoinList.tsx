import styles from './CoinList.module.scss';
import deleteIcon from '../../deleteIcon.svg';
import icon from '../../icon.svg';
import { ICoin } from '../../common/types';
import { Loader } from '../Loader/Loader';

type Props = {
  loading: boolean;
  coins: ICoin[];
  setCoins: React.Dispatch<React.SetStateAction<ICoin[]>>;
};

export const CoinList = ({ loading, coins, setCoins }: Props) => {
  if (loading) {
    return (
      <div className={styles.loading}>
        <Loader />
      </div>
    );
  }

  const deleteCoin = (id: string) => {
    const filteredCoins = coins.filter((c) => c.id !== id);
    setCoins(filteredCoins);
    localStorage.setItem('coinsInfo', JSON.stringify(filteredCoins));
  };

  return (
    <>
      {coins.map((coin, idx) => {
        return (
          <div className={styles.wrapper} key={`${coin.id}_${idx}`}>
            <div className={styles.leftSection}>
              <div>
                <img src={icon} alt="icon" />
              </div>
              <div className={styles.price}>
                <div>{coin.baseSymbol}</div>
                <div>{coin.ticker.lastPrice.substring(0, 8)} &euro;</div>
              </div>
            </div>
            <div className={styles.delete} onClick={() => deleteCoin(coin.id)}>
              <img src={deleteIcon} alt="deleteIcon" />
            </div>
          </div>
        );
      })}
    </>
  );
};
