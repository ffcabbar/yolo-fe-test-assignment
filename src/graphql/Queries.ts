import { gql } from '@apollo/client';

export const GET_COIN_PRICE_QUERY = gql`
  query price($code: String!) {
    markets(
      filter: {
        baseSymbol: { _eq: $code }
        quoteSymbol: { _eq: "EUR" }
        marketStatus: { _eq: Active }
      }
    ) {
      id
      baseSymbol
      ticker {
        lastPrice
      }
    }
  }
`;
