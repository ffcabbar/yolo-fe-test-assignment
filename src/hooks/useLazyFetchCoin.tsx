import { useLazyQuery } from '@apollo/client';
import { useEffect, useState } from 'react';
import { ICoin } from '../common/types';
import { GET_COIN_PRICE_QUERY } from '../graphql/Queries';
import { alertNotification } from '../helpers/alert-notification';
import { Notification } from '../common/enums';

export const useLazyFetchCoin = () => {
  const [coins, setCoins] = useState<ICoin[]>([]);

  useEffect(() => {
    const coins = localStorage.getItem('coinsInfo');
    if (coins) {
      setCoins(JSON.parse(coins));
    }
  }, []);

  const [getData, { loading, error }] = useLazyQuery(GET_COIN_PRICE_QUERY, {
    fetchPolicy: 'network-only',
    notifyOnNetworkStatusChange: true,
    onCompleted: (data) => {
      const hasSameCoin = coins.some((f) => f.id === data.markets[0]?.id);
      if (data.markets.length && !hasSameCoin) {
        const allCoins = [...coins, data.markets[0]];
        setCoins(allCoins);
        localStorage.setItem('coinsInfo', JSON.stringify(allCoins));
      } else if (data.markets.length <= 0) {
        alertNotification('Coin not found !', Notification.ERROR);
      }
      if (hasSameCoin) {
        alertNotification('This coin already exists on your list !', Notification.WARNING);
      }
    }
  });

  return { coins, setCoins, getData, loading, error };
};
