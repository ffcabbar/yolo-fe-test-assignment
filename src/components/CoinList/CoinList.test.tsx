import { MockedProvider } from '@apollo/react-testing';
import { renderHook, act } from '@testing-library/react-hooks';
import { useLazyFetchCoin } from '../../hooks/useLazyFetchCoin';
import { GET_COIN_PRICE_QUERY } from '../../graphql/Queries';
import { CoinList } from './CoinList';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

const TEST_VALUE = [
  {
    id: 'binance_bnb_eur',
    baseSymbol: 'BNB',
    ticker: {
      lastPrice: '414.90000000'
    }
  }
];

const QueryMock = [
  {
    request: {
      query: GET_COIN_PRICE_QUERY,
      variables: { code: 'BNB' }
    },
    result: {
      data: {
        markets: TEST_VALUE
      }
    }
  }
];

const getHookWrapper = (mocks?: any, code?: string) => {
  const wrapper = ({ children }: any) => (
    <MockedProvider mocks={mocks} addTypename={false}>
      {children}
    </MockedProvider>
  );
  const { result, waitForNextUpdate } = renderHook(() => useLazyFetchCoin(), {
    wrapper
  });

  const { coins } = result.current;
  expect(typeof coins).toBe('object');

  // call the lazy function
  act(() => {
    result.current.getData({
      variables: { code }
    });
  });

  return { result, waitForNextUpdate };
};

describe('CoinList component', () => {
  it('should not render when there is no request', async () => {
    const { result, waitForNextUpdate } = getHookWrapper();
    await waitForNextUpdate();

    const { coins, setCoins, loading } = result.current;
    const { container } = render(<CoinList coins={coins} setCoins={setCoins} loading={loading} />);
    expect(container).toBeEmptyDOMElement();
  });

  it('matches snapshot correctly when there is no request', async () => {
    const { result, waitForNextUpdate } = getHookWrapper();
    await waitForNextUpdate();

    const { coins, setCoins, loading } = result.current;
    const tree = renderer
      .create(<CoinList coins={coins} setCoins={setCoins} loading={loading} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders when there is a request', async () => {
    const { result, waitForNextUpdate } = getHookWrapper(QueryMock, 'BNB');
    await waitForNextUpdate();

    const { coins, setCoins, loading } = result.current;
    const { container } = render(<CoinList coins={coins} setCoins={setCoins} loading={loading} />);
    expect(container).toBeInTheDocument();
  });

  it('matches snapshot correctly when there is a request', async () => {
    const { result, waitForNextUpdate } = getHookWrapper(QueryMock, 'BNB');
    await waitForNextUpdate();

    const { coins, setCoins, loading } = result.current;
    const tree = renderer
      .create(<CoinList coins={coins} setCoins={setCoins} loading={loading} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
