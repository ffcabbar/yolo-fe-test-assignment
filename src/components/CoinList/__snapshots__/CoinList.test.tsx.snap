// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinList component matches snapshot correctly when there is a request 1`] = `
<div
  className="wrapper"
  data-testid="coin-list-binance_bnb_eur"
>
  <div
    className="leftSection"
  >
    <div>
      <img
        alt="icon"
        src="icon.svg"
      />
    </div>
    <div
      className="price"
    >
      <div>
        BNB
      </div>
      <div>
        414.9000
         â‚¬
      </div>
    </div>
  </div>
  <div
    className="delete"
    data-testid="coin-list-delete-binance_bnb_eur"
    onClick={[Function]}
  >
    <img
      alt="deleteIcon"
      src="deleteIcon.svg"
    />
  </div>
</div>
`;

exports[`CoinList component matches snapshot correctly when there is no request 1`] = `null`;
